cpp/src/community/detail/common_methods.cuh

```diff
@@ -212,7 +218,11 @@ graph_contraction(raft::handle_t const& handle,
     labels.size(),
     false);
```

Comment:
This only does graph contraction and return the renumber map (which are Leiden assignment of aggregate graph). Will make  `labels` read only in next commit.

Reply:
Did you make this change?

Once `labels` becomes read only,

```
  auto new_graph_view = new_graph.view();

  rmm::device_uvector<vertex_t> numbering_indices((*numbering_map).size(), handle.get_stream());
  detail::sequence_fill(handle.get_stream(),
                        numbering_indices.data(),
                        numbering_indices.size(),
                        new_graph_view.local_vertex_partition_range_first());
  relabel<vertex_t, multi_gpu>(
    handle,
    std::make_tuple(static_cast<vertex_t const*>((*numbering_map).begin()),
                    static_cast<vertex_t const*>(numbering_indices.begin())),
    new_graph_view.local_vertex_partition_range_size(),
    labels.data(),
    labels.size(),
    false);
```
is no longer necessary? Right? And this function will just call coarsen_graph, so the caller can just directly call coarsen_graph, isn't it?

Reply:
`labels `are not read only. Sorry for the confusion. 
In case of Louvain, `labels `are reference to dendrogram and  they are renumber according to new indices of contracted graph. So we can't make labels read only. 

In case of Leiden we are not using this function.  
Rather we need two relabeling

1. To get back **louvain cluster keys** of contracted nodes (where nodes are contracted based on refined leiden cluster keys)
2.  renumber the dendrogram based on number of unique **louvain cluster keys** 

I can change the function to return 
```
std::tuple<
  cugraph::graph_t<vertex_t, edge_t, false, multi_gpu>,
  std::optional<edge_property_t<graph_view_t<vertex_t, edge_t, false, multi_gpu>, weight_t>>>
```
instead of additional `numbering_map`

Reply:
Changed in last commit.